base/io.jl

```diff
@@ -475,10 +484,39 @@ read(filename::AbstractString, args...) = open(io->read(io, args...), convert(St
 read(filename::AbstractString, ::Type{T}) where {T} = open(io->read(io, T), convert(String, filename)::String)
 
 """
-    read!(stream::IO, array::AbstractArray)
-    read!(filename::AbstractString, array::AbstractArray)
+    read!(stream::IO, a::AbstractArray{T})
+    read!(stream::IO, a::Ref{T})
+    read!(filename::AbstractString, a::Union{AbstractArray{T}, Ref{T}})
+
+Read `length(a)` consecutive values of type `T` from an I/O stream or
+file into `a`, returning `a`.
+
+If `a` is an `Array` and `isbitstype(T)` is true, this function reads
+`size(a)` bytes directly into memory. Otherwise, each element of
+`a::AbstractArray{T}` is read with a call to `read(stream, T)`.
```

Comment:
```suggestion
If `isbitstype(T)` is true, this function reads `sizeof(a)` bytes
directly into memory. Otherwise, each element of `a` is populated
sequentially by a call to `read(stream, T)`.
```

Reply:
Not sure if `sizeof(a)` is accurate here.
`Base.sizeof` is not defined for most `AbstractArray` and it falls back to `Core.sizeof` by default.

BTW, if we expect `AbstractArray`s to behave like `Array`s, then the optimization for `StridedArray`s needs to check `elsize(A) == elsize(Array{eltype(A)})` to avoid possible padding mismatch.

Reply:
yeah, maybe `sizeof(T) * length(a)` bytes would be better then? (more accurate yet would be `Base.aligned_sizeof`, but that seems a bit overly pedantic to me)